core/src/main/java/com/linecorp/armeria/client/http/encoding/HttpDecodingClient.java

```diff
@@ -0,0 +1,92 @@
+/*
+ * Copyright 2017 LINE Corporation
+ *
+ * LINE Corporation licenses this file to you under the Apache License,
+ * version 2.0 (the "License"); you may not use this file except in compliance
+ * with the License. You may obtain a copy of the License at:
+ *
+ *   http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS, WITHOUT
+ * WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied. See the
+ * License for the specific language governing permissions and limitations
+ * under the License.
+ */
+
+package com.linecorp.armeria.client.http.encoding;
+
+import static com.google.common.collect.ImmutableMap.toImmutableMap;
+
+import java.util.Map;
+import java.util.function.Function;
+
+import com.google.common.collect.ImmutableList;
+import com.google.common.collect.Streams;
+
+import com.linecorp.armeria.client.Client;
+import com.linecorp.armeria.client.ClientRequestContext;
+import com.linecorp.armeria.client.DecoratingClient;
+import com.linecorp.armeria.common.http.HttpHeaderNames;
+import com.linecorp.armeria.common.http.HttpRequest;
+import com.linecorp.armeria.common.http.HttpResponse;
+
+/**
+ * A {@link DecoratingClient} that requests and decodes HTTP encoding (e.g., gzip) that has been applied to the
+ * content of a {@link HttpResponse}.
+ */
+public final class HttpDecodingClient
+        extends DecoratingClient<HttpRequest, HttpResponse, HttpRequest, HttpResponse> {
```

Comment:
Ended up with ```<I, O, I, HttpResponse``` since the type of the response for the decorator differs from the delegate. In practice, I guess a type-specific decorator chain will stop at this guy with or without the type parameters.